R/manual_layer_dag.R

```diff
@@ -86,54 +86,64 @@ DAG <- R6::R6Class(
 
     # This is important so our status display is non-frustrating for the user.
     # This assumes a cyclicity check has already been done.
-    # https://en.wikipedia.org/wiki/Topological_sorting Technically this is a
-    # reverse topological sort -- terminals appear first, then the ones that
-    # depend on them. This is intuitive for the user (and this topological sort
-    # is *only* for display to the user) since we normally write 'z <- f(x,y)'
-    # rather than 'f(x,y) -> z' -- so it makes sense to show z, then x and y.
+    #
+    # See https://en.wikipedia.org/wiki/Topological_sorting
+    #
+    # Note that this algorithm requires a destroyable copy of the graph, and it
+    # requires both forward and reverse arrows. For purpose of this function we
+    # use two R named-lists-of-named-lists as forward and reverse adjacency
+    # matrices.
     sort_nodes_topologically = function() {
       if (is.null(self$initial_nodes) || length(self$initial_nodes) == 0) {
         stop("initial nodes must be identified before topological sort")
       }
+
+      # Forward: from each arg to its calling node. E.g. if b = f(a), then row 'a' has column 'b'.
+      # Reverse: rom each node to its arg. E.g. if b = f(a), then row 'b' has column 'a'.
+      # Note forward and reverse are simply transposes of one another but it's easiest to
+      # keep two copies.
+      fwd_adj_mx <- spmx$new()
+      rev_adj_mx <- spmx$new()
+      for (node in self$all_nodes) {
+        for (arg in node$args) {
+          if (is(arg, "Node")) {
+            fwd_adj_mx$add(arg$id, node$id)
+            rev_adj_mx$add(node$id, arg$id)
+          }
+        }
+      }
+      # sorted_nodes ← Empty list that will contain the sorted elements
       sorted_nodes <- list()
+      # initial_nodes ← Set of all nodes with no incoming edge
       initial_nodes <- self$initial_nodes
 
+      # while initial_nodes is not empty do
       while (length(initial_nodes) > 0) {
+        # remove node from initial_nodes
         node <- initial_nodes[[1]]
         initial_nodes[[1]] <- NULL
 
-        # If we have the terminal first and initials last, this
-        # edge-enumeration is very quick: 'for (arg in node$args)'. For
-        # display, initials-first or terminals-first is equally nice. But for
-        # future use with server-side task graphs, we'll need initials first.
-        # So we may as well do this initials-first.
+        # add node to sorted_nodes
         sorted_nodes[[node$id]] <- node
 
-        # Initials first:
-        for (other in self$find_nodes_depending_directly_on(node)) {
-          initial_nodes[[other$id]] <- other
-        }
-      }
+        # for each node other with an edge from node to other do
```

Comment:
"for each node that is a child of other" or something like that might be a bit clearer

Reply:
lol good call. i was (modulo name changes) verbatim copy-pasting the wikipedia pseudocode i'd failed to diligently follow the first time i wrote this :^/

Reply:
(although i don't like 'child' & 'parent' since i have to disentangle those in my head as well with regard to whether arg is child or parent of node)

Reply:
"for each child that depends upon this node" is another phrasing that reduces that ambiguity. All up to you!